### ISSUE TITLE
[funcol] wait() breaks the chain for backwards

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#151240]
As reported by @lw, and captured by this repro, the wait() operation appears to 'detach' its waited output tensor and prevent gradients from flowing during backwards. The result is that any upstream paramters will not receive the expected grads.

```
import torch
import torch.distributed
import torch.distributed.tensor.parallel
torch.distributed.init_process_group(backend="nccl", rank=0, world_size=1, device_id=torch.device("cuda", 0), init_method="tcp://127.0.0.1:2743")
device_mesh = torch.distributed.device_mesh.DeviceMesh.from_group(torch.distributed.group.WORLD, "cuda")
emb = torch.nn.Embedding(128, 64).cuda()
emb = torch.distributed.tensor.parallel.parallelize_module(
   emb, device_mesh, torch.distributed.tensor.parallel.RowwiseParallel()
)
w = torch.randn(64, device="cuda", requires_grad=True)
a = emb(torch.randint(0, 128, (1024,), device="cuda"))
b = a.wait()
c = b + w
c.pow(2).sum().backward()
print(f"{a.requires_grad=}")
print(f"{b.requires_grad=}")
print(f"{emb.weight.grad is None=}")
print(f"{w.grad is None=}")
# Output:
# a.requires_grad=True
# b.requires_grad=False
# emb.weight.grad is None=True
# w.grad is None=False
```

**Comment by wconstab** [comment]
It appears this is the intended behavior since originally the design for functional collectives was to not support backwards, and indeed we have a unit test guarding that gradients are not propagated backwards.
https://github.com/pytorch/pytorch/blob/main/test/distributed/test_functional_api.py#L421

I believe it is actually not just the 'wait()' operation that breaks the chain, but even if wait passed the gradient back the collective itself would stand in the way since the collective also has no backward formula?  In any case, we have been discussing adding backwards formulae to funcol in https://github.com/pytorch/pytorch/issues/148690, and if we can do that then i'm confident we'll also fix this.

**Comment by wconstab** [comment]
Yes, I'm hoping we can just fix this by supporting backward but if not it should be feasible to error or warn.
